src/BuildingRegistry/Building/Building_State.cs

```diff
@@ -315,18 +328,34 @@ private void When(BuildingMeasurementWasChanged @event)
                 new ExtendedWkbGeometry(@event.ExtendedWkbGeometryBuilding),
                 BuildingGeometryMethod.MeasuredByGrb);
 
-            var buildingUnitPersistentLocalIds = @event.BuildingUnitPersistentLocalIds.Concat(@event.BuildingUnitPersistentLocalIdsWhichBecameDerived);
+            var buildingUnitPersistentLocalIds =
+                @event.BuildingUnitPersistentLocalIds.Concat(@event.BuildingUnitPersistentLocalIdsWhichBecameDerived);
 
             foreach (var buildingUnitPersistentLocalId in buildingUnitPersistentLocalIds)
             {
-                var buildingUnit = BuildingUnits.Single(x => x.BuildingUnitPersistentLocalId == buildingUnitPersistentLocalId);
+                var buildingUnit =
+                    BuildingUnits.Single(x => x.BuildingUnitPersistentLocalId == buildingUnitPersistentLocalId);
 
                 buildingUnit.Route(@event);
             }
 
             _lastEvent = @event;
         }
 
+        private void When(BuildingWasMerged @event)
+        {
+            BuildingStatus = BuildingStatus.Retired;
+            _lastEvent = @event;
+        }
+
+        private void When(BuildingUnitWasMoved @event)
+        {
+            var buildingUnit =
+                BuildingUnits.Single(x => x.BuildingUnitPersistentLocalId == @event.BuildingUnitPersistentLocalId);
+            _buildingUnits.Remove(buildingUnit);
+            _lastEvent = @event;
```

Comment:
I'm not sure if we should set the _lastEvent here? We only seem to do it for the BuildingUnitWasPlannedV2 and CommonBuildingUnitWasAddedV2 events. For the other building unit related events, we don't set the _lastEvent on Building_State.

Reply:
Exactly, when adding units the collection of the building gets altered so the `_lastEvent` is changed.
But this is also altering the collection by removing them.

But looking at the code, seems we have a bug.
`BuildingUnitWasRemovedV2`, `BuildingUnitRemovalWasCorrected` should also do this!

Reply:
@emalfroy will you add settings the _lastEvent for these events as well?

Reply:
Fixes in another pr